### PR TITLE
feat: add `array/base/mskbinary4d`

### LINKED ISSUE #3179: [RFC]: add `array/base/mskbinary4d`
### Description

This RFC proposes adding a new package to the `array/base/*` namespace: `@stdlib/array/base/mskbinary4d`.

This package is the 4d equivalent of [`@stdlib/array/base/mskbinary2d`](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/array/base/mskbinary2d).

To see an examples of similar packages, see [`@stdlib/array/base/mskunary3d`](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/array/base/mskunary3d) and [`@stdlib/array/base/binary4d`](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/array/base/binary4d).

While the package proposed in this RFC will **not** match the implementations found in the above packages, those packages should provide a conceptual idea of what is desired. Do **not** simply copy-paste the code found in those packages.

### Notes

- If you are interested in contributing a PR which addresses this RFC and are a **first-time contributor** or not yet familiar with our project conventions, please do **not** submit LLM-generated code. Please consult our [contributing guidelines](https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md) and the associated [development guide](https://github.com/stdlib-js/stdlib/blob/develop/docs/development.md). Failure to respect project conventions will result in your PR being rejected without review. Thank you for understanding!

### Checklist

- [X] I have read and understood the [Code of Conduct](https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md).
- [X] Searched for existing issues and pull requests.
- [X] The issue name begins with `[RFC]:`.

### PR DESCRIPTION
**Pull request by AbhijitRaut04**
Resolves #3179 .

## Description

> What is the purpose of this pull request?

This pull request:

-  implements the mskbinary4d package for masked binary operations in the array/base namespace.
- and provides a 4D equivalent of the mskbinary4d package for binary callbacks using a mask array.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #3179  

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/array/base/mskbinary4d/docs/repl.txt
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/mskbinary4d/docs/repl.txt]: This description needs updating.

File: lib/node_modules/@stdlib/array/base/mskbinary4d/docs/types/index.d.ts
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/mskbinary4d/docs/types/index.d.ts]: Description needs updating.

File: lib/node_modules/@stdlib/array/base/mskbinary4d/test/test.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/mskbinary4d/test/test.js]:
```suggestion
	shape = [ 1, 1, 2, 2 ];
	x = [
```

File: lib/node_modules/@stdlib/array/base/mskbinary4d/test/test.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/mskbinary4d/test/test.js]:
```suggestion
	mskbinary4d( [ x, y, m, z ], [ 0, 1, 2, 2 ], clbk );
```

Use consistent spacing here and below.

**Review by kgryte — COMMENTED** [review]
This is looking good. Thanks for working on this. Once the initial nitpicks have been resolved, I'll do a closer final review.

File: lib/node_modules/@stdlib/array/base/mskbinary4d/benchmark/benchmark.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/mskbinary4d/benchmark/benchmark.js]:
```suggestion
		N = floor( pow( pow( 10, i ), 1.0/4.0 ) );
```

File: lib/node_modules/@stdlib/array/base/mskbinary4d/benchmark/benchmark.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/mskbinary4d/benchmark/benchmark.js]:
```suggestion
		sh = [ N, N, N, N ];
```

File: lib/node_modules/@stdlib/array/base/mskbinary4d/benchmark/benchmark.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/mskbinary4d/benchmark/benchmark.js]:
```suggestion
		bench( pkg+'::equidimensional:size='+numel( sh ), f );
```

**Comment by kgryte** [comment]
One thing that needs to be resolved before further review is the failing unit tests.

**Comment by kgryte** [comment]
@AbhijitRaut04 One last thing, you'll want to make sure you are branching off of `develop`. Based on the commit history above, that does not appear to be the case. Your dev flow should be

```
$ git checkout develop
$ git pull upstream develop
$ git checkout -b mskbinary4d
```

Do work. Open a PR. Get it merged.

Then, to work on a new feature, repeat the above.

```
$ git checkout develop
$ git pull upstream develop
$ git checkout -b feat/new_feature
```

where `upstream` has been configured to be

```
$ git remote add upstream https://github.com/stdlib-js/stdlib.git
```

**Comment by AbhijitRaut04** [comment]
> @AbhijitRaut04 One last thing, you'll want to make sure you are branching off of `develop`. Based on the commit history above, that does not appear to be the case. Your dev flow should be
> 
> ```
> $ git checkout develop
> $ git pull upstream develop
> $ git checkout -b mskbinary4d
> ```
> 
> Do work. Open a PR. Get it merged.
> 
> Then, to work on a new feature, repeat the above.
> 
> ```
> $ git checkout develop
> $ git pull upstream develop
> $ git checkout -b feat/new_feature
> ```
> 
> where `upstream` has been configured to be
> 
> ```
> $ git remote add upstream https://github.com/stdlib-js/stdlib.git
> ```

@kgryte I have now discarded the commits of develop branch which i have raised. Will it work ?

**Comment by kgryte** [comment]
>  I have now discarded the commits of develop branch which i have raised. Will it work ?

My comment was intended to apply to all future PRs. For the current PR, you should leave the Git history, as is. The current diff is fine.

**Comment by AbhijitRaut04** [comment]
> > I have now discarded the commits of develop branch which i have raised. Will it work ?
> 
> My comment was intended to apply to all future PRs. For the current PR, you should leave the Git history, as is. The current diff is fine.
I have raised a new pr for issue #3180 after running above commands but the history is not refining.

**Comment by kgryte** [comment]
> I have raised a new pr for issue https://github.com/stdlib-js/stdlib/issues/3180 after running above commands but the history is not refining.

Likely because you have messed up your local `develop` commit history. I suggest holding off on any further PRs. Prioritize getting your current set of PRs merged. Once those PRs are merged, then you should delete your local fork and start over by cloning the repository and from then on follow the procedure outlined above.

**Comment by AbhijitRaut04** [comment]
> > I have raised a new pr for issue #3180 after running above commands but the history is not refining.
> 
> Likely because you have messed up your local `develop` commit history. I suggest holding off on any further PRs. Prioritize getting your current set of PRs merged. Once those PRs are merged, then you should delete your local fork and start over by cloning the repository and from then on follow the procedure outlined above.

Ok I will wait to for it. I have raised pr's for issues #3180 #3179 #1942 and #2106 please take a look to it. I am ready to contribute.

**Comment by AbhijitRaut04** [comment]
@kgryte I have applied the changes suggested. I kindly request you to review this PR.

**Review by kgryte — APPROVED** [review]
LGTM. Thanks, @AbhijitRaut04!

**Pull request merged**
